librad/src/git/types.rs

```diff
@@ -153,6 +153,118 @@ impl<'a> Into<ext::blob::Branch<'a>> for &'a Reference {
     }
 }
 
+pub struct SymbolicReference<'a> {
+    repo: &'a git2::Repository,
+}
+
+pub struct SymbolicReferenceSource<'a> {
+    repo: &'a git2::Repository,
+    source: String,
+}
+
+enum SymbolicReferenceTargetArgument<'a> {
+    RadReference(&'a Reference),
+    GitReference(&'a git2::Reference<'a>),
+}
+
+impl<'a> SymbolicReferenceTargetArgument<'a> {
+    pub fn resolve(&self, repo: &'a git2::Repository) -> Result<String, git2::Error> {
+        match *self {
+            SymbolicReferenceTargetArgument::RadReference(reference) => {
+                reference.find(repo)?;
+                Ok(reference.to_string())
+            },
+            SymbolicReferenceTargetArgument::GitReference(reference) => Ok(reference
+                .name()
+                .ok_or_else(|| git2::Error::from_str("Invalid target reference name"))?
+                .to_owned()),
+        }
+    }
+}
+
+pub struct SymbolicReferenceTarget<'a> {
+    repo: &'a git2::Repository,
+    source: String,
+    target: SymbolicReferenceTargetArgument<'a>,
+}
+
+pub struct SymbolicReferenceForce<'a> {
+    repo: &'a git2::Repository,
+    source: String,
+    target: SymbolicReferenceTargetArgument<'a>,
+    force_value: bool,
+}
+
+/// Builder to create a symbolic reference on the given git repo
+///
+/// Look at the subsequent `SymbolicReferenceSource`, `SymbolicReferenceTarget`
+/// and `SymbolicReferenceForce` types to see the full API.
+impl<'a> SymbolicReference<'a> {
+    /// Create the builder
+    pub fn new(repo: &'a git2::Repository) -> Self {
+        Self { repo }
+    }
+
+    /// Specify the symbolic reference source
+    pub fn source(self, source: &'a Reference) -> SymbolicReferenceSource {
```

Comment:
I think this building up of the structure is complex. I don't think we need 4 or 5 different `struct`s. We only need one that takes the arguments. Then `create` runs any of the checks. Whether we keep the `enum` for the two kinds of references or not.